statsmodels/base/_penalties.py

```diff
@@ -73,54 +86,457 @@ def grad(self, params):
-            self.wts = 1.
+    def __init__(self, weights=None, wts=None):
+        weights = _check_wts(weights, wts)  # for deprecation wts
+        if weights  is None:
+            self.weights  = 1.
         else:
-            self.wts = wts
+            self.weights  = weights
         self.alpha = 1.
 
     def func(self, params):
-        return np.sum(self.wts * self.alpha * params**2)
+        return np.sum(self.weights * self.alpha * params**2)
 
-    def grad(self, params):
-        return 2 * self.wts * self.alpha * params
+    def deriv(self, params):
+        return 2 * self.weights * self.alpha * params
+
+    def deriv2(self, params):
+        return 2 * self.weights * self.alpha * np.ones(len(params))
+
+
+class L2Univariate(Penalty):
+    """
+    The L2 (ridge) penalty applied to each parameter.
+    """
+
+    def __init__(self, weights=None):
+        if weights is None:
+            self.weights = 1.
+        else:
+            self.weights = weights
+
+    def func(self, params):
+        return self.weights * params**2
+
+    def deriv(self, params):
+        return 2 * self.weights * params
+
+    def deriv2(self, params):
+        return 2 * self.weights * np.ones(len(params))
 
 
 class PseudoHuber(Penalty):
     """
     The pseudo-Huber penalty.
     """
 
-    def __init__(self, dlt, wts=None):
+    def __init__(self, dlt, weights=None, wts=None):
         self.dlt = dlt
-        if wts is None:
-            self.wts = 1.
+        weights = _check_wts(weights, wts)  # for deprecation wts
+        if weights  is None:
+            self.weights  = 1.
         else:
-            self.wts = wts
+            self.weights  = weights
         self.alpha = 1.
 
     def func(self, params):
         v = np.sqrt(1 + (params / self.dlt)**2)
         v -= 1
         v *= self.dlt**2
-        return np.sum(self.wts * self.alpha * v)
+        return np.sum(self.weights * self.alpha * v, 0)
 
-    def grad(self, params):
+    def deriv(self, params):
         v = np.sqrt(1 + (params / self.dlt)**2)
-        return params * self.wts * self.alpha / v
+        return params * self.weights * self.alpha / v
+
+    def deriv2(self, params):
+        v = np.power(1 + (params / self.dlt)**2, -3/2)
+        return self.weights * self.alpha * v
+
+
+class SCAD(Penalty):
+    """
+    The SCAD penalty of Fan and Li.
+
+    The SCAD penalty is linear around zero as a L1 penalty up to threshold tau.
+    The SCAD penalty is constant for values larger than c*tau.
+    The middle segment is quadratic and connect the two segments with a continuous
+    derivative.
+    The penalty is symmetric around zero.
+
+    Parameterization follows Boo, Johnson, Li and Tan 2011.
+    Fan and Li use lambda instead of tau, and a instead of c. Fan and Li
+    recommend setting c=3.7.
+
+    f(x) = { tau |x|                                   if 0 <= |x| < tau
+           { -(|x|^2 - 2 c tau |x| + tau^2) / (2 (c - 1)   if tau <= |x| < c tau
```

Comment:
Missing close-paren in the denominator?  Is there a constraint `c > 1`?

Reply:
c is a parameter of the penalty function, we need tau < c * tau